iceberg/streams/iceberg/file_reader_builder.cpp

```diff
@@ -0,0 +1,92 @@
+#include "iceberg/streams/iceberg/file_reader_builder.h"
+
+namespace iceberg {
+
+namespace {
+
+void InsertOrFail(std::unordered_map<int, int>& mp, int key, int value) {
+  Ensure(mp.insert(std::make_pair(key, value)).second, std::string(__PRETTY_FUNCTION__) + ": field ids are not unique");
+}
+
+std::unordered_map<int, int> MakeFieldIdToIndexMap(const parquet::schema::GroupNode* node,
+                                                   const std::optional<SchemaNameMapper>& schema_name_mapper) {
+  std::unordered_map<int, int> field_id_to_index;
+  for (int i = 0; i < node->field_count(); ++i) {
+    auto field = node->field(i);
+    Ensure(field != nullptr, std::string(__PRETTY_FUNCTION__) + ": field is nullptr");
+    auto field_id = field->field_id();
+
+    if (field_id >= 0) {
+      InsertOrFail(field_id_to_index, field_id, i);
+      continue;
+    }
+
+    if (!schema_name_mapper.has_value()) continue;
+    auto field_name = field->name();
+    auto possible_field_id = schema_name_mapper->GetRootMapper().GetFieldIdByName(field_name);
```

Comment:
It's too expensive cycle inside. We could create a hash_map in `schema_name_mapper` to avoid it. Just extract all names from doc into hash_map once we create `schema_name_mapper`. And we get O(1) here.